src/VecSim/algorithms/hnsw/hnsw_tiered.h

```diff
@@ -619,10 +667,9 @@ int TieredHNSWIndex<DataType, DistType>::addVector(const void *blob, labelType l
     // (in case of override in non-MULTI index) - so if it's there, we remove it (and create the
     // required repair jobs), *before* we submit the insert job.
     if (!this->backendIndex->isMultiValue()) {
-        this->mainIndexGuard.lock_shared();
-        this->deleteLabelFromHNSW(label);
-        this->mainIndexGuard.unlock_shared();
-        ret = 0;
+        // If we removed the previous vector from both HNSW and flat in the overwrite process,
+        // we still return 0 (not -1).
+        ret = MAX(ret - this->deleteLabelFromHNSW(label), 0);
```

Comment:
do we have a test for this scenario?

Reply:
Yes it's part of the `bufferLimitAsync` test